src/FSharp.Core/Random.fs

```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace Microsoft.FSharp.Core
+
+open System
+open System.Threading
+
+[<AbstractClass; Sealed>]
+type internal ThreadSafeRandom() =
+    [<DefaultValue>]
+    static val mutable private globalRandom: Random
+
+    [<DefaultValue>]
+    static val mutable private localRandom: ThreadLocal<Random>
+
+    static do ThreadSafeRandom.globalRandom <- Random()
+
+    static do
+        ThreadSafeRandom.localRandom <-
+            new ThreadLocal<Random>(fun () ->
+                lock ThreadSafeRandom.globalRandom (fun () -> Random(ThreadSafeRandom.globalRandom.Next())))
+    // Don't pass the returned Random object between threads
+    static member Shared = ThreadSafeRandom.localRandom.Value
```

Comment:
@brianrourkeboll No, no particular reason I took the algorithm from SO https://stackoverflow.com/a/57962385/1780648 (looked cleaner for me). The BCL algorith can't be taken 1 in 1 since XoshiroImpl is not available and constructor with isThreadSafeRandom are not available.

Reply:
Right, but could you not still bind the `Random` instance to a field with `ThreadStaticAttribute` instead of using a `ThreadLocal` and locking? You wouldn't get the devirtualization that they get in the BCL by directly delegating to the `XoshiroImpl`, but the comment makes it sound like `ThreadStatic` might be better for concurrent access overall:

```csharp
// …we need the instance
// returned from Shared itself to be thread-safe, which can be accomplished either by locking around all
// accesses or by itself accessing a ThreadStatic on every access: we've chosen the latter, as it is more
// scalable.
```

Something like:

```fsharp
open System
open System.Runtime.CompilerServices

[<Sealed>]
type internal ThreadSafeRandom() =
    inherit Random()

    [<DefaultValue>]
    [<ThreadStatic>]
    static val mutable private random: Random

    [<MethodImpl(MethodImplOptions.NoInlining)>]
    static member private Create() =
        ThreadSafeRandom.random <- Random()
        ThreadSafeRandom.random

    static member private LocalRandom =
        match ThreadSafeRandom.random with
        | null -> ThreadSafeRandom.Create()
        | random -> random

    override _.Next() = ThreadSafeRandom.LocalRandom.Next()
    override _.Next maxValue = ThreadSafeRandom.LocalRandom.Next maxValue
    override _.Next(minValue, maxValue) = ThreadSafeRandom.LocalRandom.Next(minValue, maxValue)
    override _.NextInt64() = ThreadSafeRandom.LocalRandom.NextInt64()
    override _.NextInt64 maxValue = ThreadSafeRandom.LocalRandom.NextInt64 maxValue
    override _.NextInt64(minValue, maxValue) = ThreadSafeRandom.LocalRandom.NextInt64(minValue, maxValue)
    override _.NextSingle() = ThreadSafeRandom.LocalRandom.NextSingle()
    override _.NextDouble() = ThreadSafeRandom.LocalRandom.NextDouble()
    override _.NextBytes(buffer: byte array) = ThreadSafeRandom.LocalRandom.NextBytes buffer
    override _.NextBytes(buffer: Span<byte>) = ThreadSafeRandom.LocalRandom.NextBytes buffer
    override _.Sample() = raise (NotSupportedException())

    static member val Shared: Random = ThreadSafeRandom()

type Random with
    static member internal Shared = ThreadSafeRandom.Shared
```

That would also let you just write `Random.Shared` in FSharp.Core.

But maybe I'm missing something, or maybe the performance benefit wouldn't exist or be worth it in this case.

Reply:
All right, here's a quick [benchmark](https://gist.github.com/brianrourkeboll/438f9c53dc8d1b4eba168bbcf68b2a2a):

```console
| Method       | Mean      | Error     | StdDev    | Ratio | RatioSD |
|------------- |----------:|----------:|----------:|------:|--------:|
| Bcl          |  2.887 ns | 0.0083 ns | 0.0073 ns |  1.00 |    0.00 |
| ThreadStatic |  4.111 ns | 0.0396 ns | 0.0351 ns |  1.42 |    0.01 |
| ThreadLocal  | 10.835 ns | 0.0517 ns | 0.0404 ns |  3.75 |    0.02 |
```

It looks like it does make a noticeable difference.

Reply:
Thanks for the investigation! I'll update the implementation

Reply:
I've updated to the simplified version (similar to what I had with ThreadLocal), since it shows a bit better performance

```
| Method             | Mean     | Error     | StdDev    | Median   | Ratio | RatioSD | Allocated | Alloc Ratio |
|------------------- |---------:|----------:|----------:|---------:|------:|--------:|----------:|------------:|
| Bcl                | 4.006 ns | 0.1068 ns | 0.1631 ns | 3.919 ns |  1.00 |    0.00 |         - |          NA |
| ThreadStatic       | 5.529 ns | 0.1308 ns | 0.1701 ns | 5.462 ns |  1.37 |    0.07 |         - |          NA |
| ThreadStaticSimple | 4.706 ns | 0.0677 ns | 0.0633 ns | 4.686 ns |  1.15 |    0.06 |         - |          NA |
```

Reply:
Hmm, I don't that works, though. I believe you need the single wrapper instance with a thread-static field.

In the version in 82860d8, `Shared` will publicly return a separate `Random` object per thread — each of which _could_ then be accessed from other threads, which would _not_ be thread-safe.

In the way I suggested based on the BCL, there is only ever one `Shared` instance of `Random`, namely the single instance of `ThreadSafeRandom`. That instance in turn delegates calls to private per-thread `Random` instances, each of which will never be accessed from or passed to another thread.

Try this out: https://gist.github.com/brianrourkeboll/5aa6d786f716995c53cca31cdca90e54

Reply:
Yes, I understand that, except that it is `internal`, not `public`, so there is no risk that end user misuse it. Current implementation is safe, it can only become unsafe if someone want to implement something like "Array.Parallel.randomShuffle" in Fsharp.Core, which is highly unlikely (I still added a warning comment). I can change the implementation for the future-proof one, if you insist, but I think that having a simple and a bit more performant version is beneficial.

Reply:
I see what you mean. It may be worth picking other names than `ThreadSafeRandom` and `Shared`, then, since the ones here aren't thread-safe or shared in the same way that the BCL ones with those names are.

Reply:
Ok, instead of `ThreadSafeRandom.Shared` I can suggest `ThreadLocalRandom.Instance`

Reply:
Since this is internal, we can probably change this later in case further ambiguity. We can keep it like this for now.